WP Scripts: Automatically detect entry points for the default JS and CSS files

gziolo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #55936.
Related to #65602.

Improves the default webpack configuration to become more flexible in how entry points get detected. Primarily, make it possible for the main CSS files get detected in the source directory and set as an entry point. Secondly, make it more likely that the JavaScript main file is set as an entry point if it isn't directly tied to any block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

It's been reported multiple times that WP Scripts is too much tied to block development. The expectation it works out of the box when writing JavaScript for plugins or styles for themes. The good example is the latest blog post from @justintadlock on WordPress Developer Blog: [Getting and setting Block Binding values in the Editor](https://developer.wordpress.org/news/2024/10/29/getting-and-setting-block-binding-values-in-the-editor/). In particular, this PR tries to mitigate the need for a custom webpack config in the project like:

```js
// WordPress webpack config.
const defaultConfig = require( '@wordpress/scripts/config/webpack.config' );

// Utilities.
const path = require( 'path' );

// Add any a new entry point by extending the webpack config.
module.exports = {
	...defaultConfig,
	...{
		entry: {
			'js/editor': path.resolve( process.cwd(), 'resources/js', 'editor.js' )
		}
	}
};
```

Taking into account  the project structure explained in the article:

<img width="236" alt="Screenshot 2024-11-10 at 11 39 40" src="https://github.com/user-attachments/assets/6fab97ce-90e1-48f1-8636-14eb364c2ba8">

With the changes proposed, the entry point could still be located in the `resources/js/` folder, but the file would need to get renamed to `index.js`. Then the build command would become:

```bash
wp-scripts build --webpack-src-dir=resources/js --output-path=public
```

This would altogether remove the need for the custom config file.

_An open question is whether we should also handle different names as entry points in addition to `index.js`. In this case, `editor.js` could be the alternative worth considering._

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

It is based on the convention where JS file is `index.js` (following Node.js) and the CSS file is `style.css` ([main stylesheet in themes](https://developer.wordpress.org/themes/basics/main-stylesheet-style-css/)).

Detection logic for the list of entry points with the updated webpack config. There are several steps:
1. Use the legacy webpack 4 format passed as CLI arguments, example: `wp-scripts build index.js`. Using this option will set entry points based on the list of the paths passed.
2. Scan `block.json` files for scripts listed inside `editorScript`, `script` and `viewScript` fields. If `block.json` is located in the root of the source directory, the scanning ends for backward compatibility.
3. Adds `src/index.*` JavaScript file when exists.
4. Adds `src/style.*` CSS file when exists.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

```bash
npx wp-create-block example-block --no-wp-scripts
cd example-block
../node_modules/.bin/wp-scripts build
```

This covers the single block example in which case, steps (3) and (4) outlined in the "How?" section should be skipped because `block.json` is located in the root of the source directory.

<img width="410" alt="Screenshot 2024-11-10 at 14 25 16" src="https://github.com/user-attachments/assets/385ddb4d-fef3-44b2-b3e4-160b92c75e86">

The `build` folder looks as follows:

<img width="442" alt="Screenshot 2024-11-10 at 14 26 06" src="https://github.com/user-attachments/assets/fd003138-4919-4c49-9ae4-e1e0e15aa66a">

## Documentation changes

The above-mentioned and other possible configurations are explained in [the updated README file](https://github.com/WordPress/gutenberg/blob/7ce56960768287e833a70b76c2d6eb8788c69bbc/packages/scripts/README.md), too.

gziolo committed:
Scripts: Automatically detect entry points for the default JS and CSS files

gziolo commented on packages/scripts/utils/config.js:
```diff
@@ -309,42 +311,53 @@ function getWebpackEntryPoints( buildType ) {
-			absolute: true,
-			cwd: fromProjectRoot( getWordPressSrcDirectory() ),
-		} );
+		const hasBlockJsonInRoot = blockMetadataFiles.includes(
```
This check is broken on Windows. I will investigate.

gziolo commented on packages/scripts/utils/config.js:
```diff
@@ -281,7 +283,7 @@ function getWebpackEntryPoints( buildType ) {
 							)
 						);
-						return;
+						continue;
```
@sirreal, I checked the previous implementation before the function has support for `buildType` added in https://github.com/WordPress/gutenberg/pull/57461 and the `return` was inside the callback rather than loop. So it should continue execution and skip only misconfigured entries.

sirreal replied:
I think you're correct, these `continue`s look like bugfixes.

Let's extract this change to its own PR.

gziolo replied:
Excellent idea. While I played with it, I discovered more issues that I fixed in https://github.com/WordPress/gutenberg/pull/66752.

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -121,7 +122,9 @@ const baseConfig = {
 					test: /[\\/]style(\.module)?\.(pc|sc|sa|c)ss$/,
-					chunks: 'all',
+					chunks( chunk ) {
+						return chunk.name !== 'style';
```
```suggestion
						// Ensure that the main style file can still be generated when provided as an entry point.
						return chunk.name !== 'style';
```

gziolo committed:
Update packages/scripts/config/webpack.config.js

gziolo committed:
Improve documentation for the default webpack config

gziolo committed:
Improve code comments

gziolo committed:
Fix the file check broken on Windows

gziolo committed:
Improve code documentation

gziolo committed:
Improve the documentation for the project structure and build scripts

gziolo committed:
Add changelog entry

gziolo commented:
I included the changelog entry with https://github.com/WordPress/gutenberg/pull/66466/commits/7ce56960768287e833a70b76c2d6eb8788c69bbc. I marked changes as breaking as it's very hard to predict whether all existing projects will work precisely the same. However, I put as much care as I could into understanding the implications to minimize the risk.

youknowriad commented:
Sorry if my question is out of line, I kind of lost track about how all these tools work over time. But why does `wp-scripts build` work on a "directory" as entry point instead of a file (most recent bundlers use files as entry points). There are probably valid reason, I guess some convention that we set up a long time ago.

So here's a proposal that may or may not be valid depending on the answer to the question above but what if we change that and require an entry point to the build command:

```
wp-scripts build src/index.js
wp-scripts build assets/style.css
wp-scripts build blocks/block.json
```

What use-cases we would be missing in this case?

gziolo commented:
@youknowriad, great questions. I must admit that given how long the WP Scripts tool is around, I also don’t fully remember which conventions we established in which order and when we over optimized for plugins with blocks 😅

What’s interesting webpack 4 supported the syntax you mentioned:

```bash
wp-scripts build index.js
```

That’s why we added a compatibility layer on top of webpack 5 to retain the functionality. Unfortunately, it doesn’t work with CSS or block metadata. What’s interesting, when you pass as many paths to files as you want they are used as entry points and we don’t use any automatic discovery for anything else. What would be compatible with your perspective is extending this format to support more asset types. Example:

```bash
wp-scripts build index.ts style.scss **/block.json
```

This is more or less the equivalent of the default behavior that I was trying to cover with the changes proposed. However, I’m happy to explore the alternative where you can pass paths to JS, CSS and block metadata explicitly of with globs that are further processed by the bundler and converted into entry points based on existing rules.

The most complex setup I have seen we would need to potentially cover is multiple plugins with multiple blocks, and  multiple themes in one project. That’s rather something that requires a custom configuration, but the approach you shared would allow a lot of flexibility in collecting all necessary entry points by listing files in the command.

sirreal commented on packages/scripts/config/webpack.config.js:
```diff
@@ -377,6 +382,10 @@ const scriptConfig = {
 		! process.env.WP_NO_EXTERNALS &&
 			new DependencyExtractionWebpackPlugin(),
+		// Removes obsolete empty scripts generated when CSS files are set as entry points.
+		new RemoveEmptyScriptsPlugin( {
+			stage: RemoveEmptyScriptsPlugin.STAGE_AFTER_PROCESS_PLUGINS,
+		} ),
```
❤️ 
> RemoveEmptyScriptsPlugin.STAGE_AFTER_PROCESS_PLUGINS
Remove empty scripts after processing all other plugins.
For example, exact this stage needs for properly work of the @wordpress/dependency-extraction-webpack-plugin.

sirreal commented on packages/scripts/utils/config.js:
```diff
@@ -309,35 +312,49 @@ function getWebpackEntryPoints( buildType ) {
+		if ( ! hasBlockJsonInRoot ) {
+			// 3. Checks whether a standard JavaScript file name can be detected in the defined source directory,
+			//    and converts the discovered file to an entry point.
+			const [ entryJavaScriptFile ] = glob( 'index.?(m)[jt]s?(x)', {
+				absolute: true,
+				cwd: fromProjectRoot( getWordPressSrcDirectory() ),
```
I'd support the `c` (commonJS) prefix here, but I'd probably remove and reserve `m` (module) prefix as a script-modules indicator. What do you think?

Related: #65271

gziolo replied:
Sounds good to me 👍

sirreal reviewed (approved):
This looks good, I left a couple notes, specifically I'd avoid supporting extensions lik `.mjs` right now as that would be a nice quick way to support modules.

youknowriad commented:
> That’s why we added a compatibility layer on top of webpack 5 to retain the functionality. Unfortunately, it doesn’t work with CSS or block metadata. What’s interesting, when you pass as many paths to files as you want they are used as entry points and we don’t use any automatic discovery for anything else. What would be compatible with your perspective is extending this format to support more asset types.

Yes, I think having a compatibility layer is fine as for me webpack is just an implementation details. I've never liked the fact that you can pass your webpack config and would prefer if we remove it entirely at some point but I guess it's ok to keep for now as it doesn't prevent us from exploring the possibility to pass any random entry point.

I'd love to avoid any `--webpack-*` something config.

gziolo commented on packages/scripts/utils/config.js:
```diff
@@ -309,35 +312,49 @@ function getWebpackEntryPoints( buildType ) {
+		if ( ! hasBlockJsonInRoot ) {
+			// 3. Checks whether a standard JavaScript file name can be detected in the defined source directory,
+			//    and converts the discovered file to an entry point.
+			const [ entryJavaScriptFile ] = glob( 'index.?(m)[jt]s?(x)', {
```
```suggestion
			const [ entryJavaScriptFile ] = glob( 'index.?(c)[jt]s?(x)', {
```

@sirreal, is what you had in mind when commenting [here](https://github.com/WordPress/gutenberg/pull/66466/files#r1842227588)?

gziolo replied:
I think we would also have to extend the support for CJS in the place where the loaders get configured:

https://github.com/WordPress/gutenberg/blob/7ce56960768287e833a70b76c2d6eb8788c69bbc/packages/scripts/config/webpack.config.js#L162

sirreal replied:
Yes, that looks like what I had in mind. If we don't want to tackle all the changes to support `.cjs` etc. I think it would be fine to just remove the `m` part from auto-discovery now.

gziolo commented:
@youknowriad, I'd like to ensure we have some shared understanding of the general direction. We should aim towards making the `wp-scripts build` configurable through the CLI params that with time become entirely independent from webpack allowing devs to define as many JS and CSS files as necessary as entry points. In addition to that, they should be able to pass paths to `block.json` files that would copy these files to the build folder, and JS and CSS files listed in metadata as paths should also get detected as entry points. CLI params that allow defining the source folder or the target build folder should also become completely independent from how webpack works. 

A more comprehensive example where we could go:

```bash
wp-scripts build index.ts editor.js style.scss **/block.json --source-path assets --output-path public
```

---

How do you feel about the changes included in this PR? Should we leave it as and land it despite the minimal risk that some existing projects extending the default webpack config could be impacted when using the `wp-scripts build` and might see two new entry points in the output directory if they used the `index.js` and `style.css` in the source directory?

youknowriad commented:
```
wp-scripts build index.ts editor.js style.scss **/block.json --source-path assets --output-path public
```

In this example, I don't know why we need "source-path" at all? But yes, your summary makes a lot of sense to me.

> How do you feel about the changes included in this PR? Should we leave it as and land it despite the minimal risk that some existing projects extending the default webpack config could be impacted when using the wp-scripts build and might see two new entry points in the output directory if they used the index.js and style.css in the source directory?

For me, if we strive to solve the former (the example you shared above), the current PR might be largely unnecessary. That said, if you feel like it's a good temporary solution, I don't mind it knowing that people will have to change later.

For the breaking changes, I think it's ok as long as the upgrade path is clear.

gziolo commented:
Excellent feedback, @youknowriad. Much appreciated. In the next few days, I will try to collect more feedback from folks actually using WP Scripts before deciding whether land these changes as the intermediate step or whether to go straight into the direction discussed above which will require more work.

> In this example, I don't know why we need "source-path" at all?

We made some assumptions on how the project is structured in the past when using `wp-scripts build` (no file paths passed) on on the default Create Block template:
- Source code lives in `src` directory.
- Build folder lands in `build` directory. 

These CLI args offer a way to change these defaults. However, after checking the variant when you explicitly pass the paths to JS files it indeed isn't necessary to provide the source folder. The only downside is 
 when the dev has the following command:

```bash
wp-scripts build src/assets/js/index.js
```

In that case, the build file would be put inside `build/src/assets/js/index.js`.

sirreal commented:
I agree with the direction that's described in the last few comments here.

I find abstraction over webpack is awkward in this tool, the implementation details related to webpack leak through in many places, yet webpack isn't fully exposed either. This makes it difficult to iterate and experiment, for example with script modules where non-webpack-based tooling may provide better alternatives.

One question I do have is regarding script modules. Right now, entrypoints can be passed as positional arguments and the tool knows how to do the right thing. This works with classic scripts, but there's no way to pass a JavaScript entrypoint that should be compiled as a script module. I don't have a great solution, I think we'll need to support an additional named argument that adds script module entrypoints. I hinted at using `.mjs`-like extensions, but I don't think that's the best solution and the extension will have other implications. _We don't need to have that sorted out in this PR_ but it's something to keep in mind. @ryanwelcher [brought this up recently in Core Slack](https://wordpress.slack.com/archives/C071CRKGKUP/p1730916550974499).

gziolo commented:
I was looking at usage examples of `build` and `start` through the lens of removing the dependency on webpack referenced.  These what is included in the docs:

```json
{
	"scripts": {
		"build": "wp-scripts build",
		"build:custom": "wp-scripts build entry-one.js entry-two.js --output-path=custom",
		"build:copy-php": "wp-scripts build --webpack-copy-php",
		"build:custom-directory": "wp-scripts build --webpack-src-dir=custom-directory"
	}
}
```

What I found interesting is that the direction discussed is fully backward compatible, and it offers a unified interface that could also work with PHP files. In effect, `build:copy-php` could get refactored and `build:custom-directory`:

```json
{
	"scripts": {
		"build": "wp-scripts build index.js **/*.php --source-path=src --output-path=custom",
	}
}
```

@sirreal, I don't have a formed vision for improved script module support. It's definitely something that we have to explore as part of these efforts. The only way we can precisely assume script modules today is through `viewScriptModule` detected in the `block.json` file. The file extension like `*.mjs` could be another indicator. Defining `type: module` in the `package.json` file, could be another way to tell the bundler to treat the passed JavaScript entry points differently.

> I find abstraction over webpack is awkward in this tool, the implementation details related to webpack leak through in many places, yet webpack isn't fully exposed either. This makes it difficult to iterate and experiment, for example with script modules where non-webpack-based tooling may provide better alternatives.

Yes, we should operate with the assumption that webpack is going to be replaced with [Rspack](https://rspack.dev), [Vite](https://vite.dev), or something else in the near future.

youknowriad commented:
For scripts modules, we should also check how the bundlers do it. Most of them offer a `--format` argument or something like that to change the format of the output.

gziolo commented:
I talked to @ndiego and @justintadlock about the idea proposed by @youknowriad in https://github.com/WordPress/gutenberg/pull/66466#issuecomment-2475822711. They shared examples of custom webpack configs they used in projects. I validated how much the new CLI syntax could remove the need to deal with webpack internals, and I got very positively surprised how quickly I could map it to the idea discussed. Here is a summary.

## Blog post example 

This is how the revised command could look for the config overrides included in the blog post [Getting and setting Block Binding values in the Edito](https://developer.wordpress.org/news/2024/10/29/getting-and-setting-block-binding-values-in-the-editor/).

```bash
npx wp-scripts build ./js/editor.js --source-path=resources --output-path=public
```

I was also surprised to learn that after a few tries, I was able to run the command with the current version of WP Scripts with:

```bash
npx wp-scripts build js/editor=./resources/js/editor.js --output-path=public
```

Note: I tried also the following `npx wp-scripts build editor.js --webpack-src-dir=resources/js --output-path=public/js`. However, `--webpack-src-dir` isn't integrated with the syntax for file paths, they are all relative to the current working directory.

## Social Sharing block example

For the [Social Sharing block](https://github.com/ndiego/social-sharing-block/blob/main/webpack.config.js), the command in the future would look as follows:

```bash
npx wp-scripts build ./*/block.json ./*/utils.php --source-path=src
```

Note: All paths by default would get resolved from the current working directory, so `src` needs to be explicitly passed to avoid the case where they end up inside build/src/*.

As of today, the following should work, too:

```bash
npx wp-scripts build --webpack-copy-php
```

Note: It would ensure all PHP files get copied from `src` to `build` directory. Both are considered default settings today in the default webpack config when no files are listed. `block.json` files get automatically detected and related JS set as entry points.

## Block Visibility plugin

Another config is for [Block Visibility](https://github.com/ndiego/block-visibility/blob/main/webpack.config.js). In the future, it could be handled with the following:

```bash
npx wp-scripts build block-visibility-editor=./src/editor.js block-visibility-setting=./src/settings.js block-visibility-editor-styles=./src/styles/editor.scss block-visibility-contextual-indicator-styles=./src/styles/contextual-indicators.scss block-visibility-setting-styles=./src/styles/settings.scss
```

The custom config is required today, because there is no support for CSS as entry points. This PR adds this concept. I'm not entirely sure how the file path syntax fits here. It might even work out of the box for the syntax provided above. It won't work without small tweaks for a simplified version like `npx wp-scripts build ./editor.css because it assumes the extension is `.js`.

## X3P0: Ideas theme

A more real-world example of what is necessary to do with themes. Here’s the [webpack.config.js]( https://github.com/x3p0-dev/x3p0-ideas/blob/master/webpack.config.js). This one is very complex as it also handles fonts and media. Something like that might work one day:

```bash
npm run build css/blocks/*/*=scss/blocks/*/*.scss js/editor=./js/editor.js css/*=./scss/*.scss {media,fonts}/*.* --source-path=resources
```

Important note: I don’t know how far we can go with these globs. The challenge is mostly for the part that maps the paths to entry points.

gziolo commented:
I decided to switch this PR to the draft state for now. Based on my most recent comment, I will work on smaller tasks to improve the existing CLI command instead of trying to improve the automatic detection for entry points.

youknowriad commented:
I would love to understand more why you had to use `source-path` argument as ideally we shouldn't need it.

gziolo commented:
> I would love to understand more why you had to use source-path argument as ideally we shouldn't need it.

Yes, it's optional. I used it to make the paths shorter 😅 I updated my comment to address your concern. We shouldn't need to promote it so heavily as it might be confusing. Still, it might be a handful option in some cases.

gziolo commented:
I started a new PR that will supersede this one. The first step is going to be recommending passing JavaScript paths as entry points:
- https://github.com/WordPress/gutenberg/pull/68251
